vm/src/obj/objlist.rs

```diff
@@ -181,22 +181,163 @@ impl PyListRef {
+            let stop = &slice.stop.as_ref().map(|x| x + 1);
+            let range = self.get_slice_range(&stop, &start);
+            match (-step).to_i32() {
+                Some(num) => self._set_stepped_slice_reverse(range, num as usize, sec, vm),
+                None => {
+                    // not sure how this is reached, step too big for i32?
+                    // then step is bigger than the than len of the list no question
+                    self._set_stepped_slice_reverse(range.end - 1..range.end, 1, sec, vm)
+                }
+            }
+        }
+    }
+
+    fn _set_slice(self, range: Range<usize>, sec: PyObjectRef, vm: &VirtualMachine) -> PyResult {
+        // consume the iter, we don't need it's size
+        // but if it's going to fail we want that to happen *before* we start modifing
+        if let Ok(items) = vm.extract_elements(&sec) {
+            // replace the range of elements with the full sequence
+            self.elements.borrow_mut().splice(range, items);
+
+            Ok(vm.get_none())
+        } else {
+            Err(vm.new_type_error("can only assign an iterable to a slice".to_string()))
+        }
+    }
+
+    fn _set_stepped_slice(
+        self,
+        range: Range<usize>,
+        step: usize,
+        sec: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult {
+        // consume the iter, we  need it's size
+        // and if it's going to fail we want that to happen *before* we start modifing
+        let slicelen = ((range.end - range.start - 1) / step) + 1;
+        if let Ok(items) = vm.extract_elements(&sec) {
+            let n = items.len();
+
+            if range.start < range.end {
+                if n == slicelen {
+                    let indexes = range.step_by(step);
+                    self._replace_indexes(indexes, &items);
+                    Ok(vm.get_none())
+                } else {
+                    Err(vm.new_value_error(format!(
+                        "attempt to assign sequence of size {} to extended slice of size {}",
+                        n, slicelen
+                    )))
+                }
+            } else {
+                // empty slice but this is an error because stepped slice
+                Err(vm.new_value_error(format!(
+                    "attempt to assign sequence of size {} to extended slice of size 0",
+                    n
+                )))
+            }
+        } else {
+            Err(vm.new_type_error("can only assign an iterable to a slice".to_string()))
+        }
+    }
+
+    fn _set_stepped_slice_reverse(
+        self,
+        range: Range<usize>,
+        step: usize,
+        sec: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult {
+        // consume the iter, we  need it's size
+        // and if it's going to fail we want that to happen *before* we start modifing
+        let slicelen = ((range.end - range.start - 1) / step) + 1;
+        if let Ok(items) = vm.extract_elements(&sec) {
```

Comment:
Can these extracts be moved outwards to ```set_slice```? (I wonder if ```set_slice``` could take a ```PyIterable```).

Reply:
I could move the check up to `setslice` and use `PyIterable` but I will have to patch `PyIterable::try_from_object` to correctly transform sequences. right now it doesn't work on strings and classes with a `__getitem__` as it should.

I actually already have a patch that does that but I'm not sure it would be a good idea to push it on this branch. thoughts?

Reply:
for reference
```
class CGetItem:
  def __init__(self, sec=(1, 2, 3)):
    self.sec = sec
  def __getitem__(self, sub):
    return self.sec[sub]

x = list(range(10))
x[3:8] = CGetItem()
assert x == [0, 1, 2, 1, 2, 3, 8, 9]
```

will work in python but `PyIterable` needs to looks somthing like
```rust
impl<T> TryFromObject for PyIterable<T>
where
    T: TryFromObject,
{
    fn try_from_object(vm: &VirtualMachine, obj: PyObjectRef) -> PyResult<Self> {
        if let Ok(method) = vm.get_method(obj.clone(), "__iter__") {
            Ok(PyIterable {
                method: method,
                _item: std::marker::PhantomData,
            })
        } else if vm.get_method(obj.clone(), "__getitem__").is_ok() {
            Self::try_from_object(
                vm,
                objiter::PySequenceIterator {
                    position: Cell::new(0),
                    obj: obj.clone(),
                }
                .into_ref(vm)
                .into_object(),
            )
        } else {
            Err(vm.new_type_error(format!("'{}' object is not iterable", obj.class().name)))
        }
    }
}
```

Reply:
Weird. I did not know that was possible. Especially as list and tuple do implement ```__iter__```. This exists to optimise strings I guess?

Either here or separate PR is fine. Should I merge this and you can continue this elsewhere?

Reply:
I actually just found a bug related to the same error as #746 so I'm going to push some more commits here shortly that will include that fix and the `try_from_object` patch

Reply:
and done, should be good now.